src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/create/actions/EmitUpdateRulesRequestsAction.java

```diff
@@ -0,0 +1,134 @@
+    }
+
+    @Override
+    protected void perform(State from, State to,
+                           Event event, FlowMirrorPointCreateContext context, FlowMirrorPointCreateFsm stateMachine) {
+        stateMachine.getCommands().clear();
+        stateMachine.getPendingCommands().clear();
+
+        String flowId = stateMachine.getFlowId();
+        Flow flow = getFlow(flowId);
+
+        FlowCommandBuilder commandBuilder = commandBuilderFactory.getBuilder(flow.getEncapsulationType());
+
+        Collection<FlowSegmentRequestFactory> commands = buildCommands(commandBuilder, stateMachine, flow);
+
+        // emitting
+        PathId flowPathId = stateMachine.getFlowPathId();
+        SwitchId mirrorSwitchId = stateMachine.getMirrorSwitchId();
+        FlowMirrorPoints mirrorPoints = flowMirrorPointsRepository.findByPathIdAndSwitchId(flowPathId, mirrorSwitchId)
+                .orElse(null);
+
+        SpeakerRequestEmitter requestEmitter;
+        if (mirrorPoints != null && mirrorPoints.getMirrorPaths().isEmpty()) {
+            requestEmitter = SpeakerRemoveSegmentEmitter.INSTANCE;
+        } else {
+            requestEmitter = SpeakerInstallSegmentEmitter.INSTANCE;
+        }
+
+        requestEmitter.emitBatch(stateMachine.getCarrier(), commands, stateMachine.getCommands());
+        stateMachine.getCommands().forEach(
+                (key, value) -> stateMachine.getPendingCommands().put(key, value.getSwitchId()));
+
+        if (commands.isEmpty()) {
+            stateMachine.saveActionToHistory("No need to update rules");
+        } else {
+            stateMachine.saveActionToHistory("Commands for updating rules have been sent");
+            stateMachine.setRulesInstalled(true);
+        }
+    }
+
+    private Collection<FlowSegmentRequestFactory> buildCommands(FlowCommandBuilder commandBuilder,
+                                                                FlowMirrorPointCreateFsm stateMachine, Flow flow) {
+        PathId flowPathId = stateMachine.getFlowPathId();
+        FlowPath path = flow.getPath(flowPathId).orElseThrow(() -> new FlowProcessingException(ErrorType.NOT_FOUND,
+                format("Flow path %s not found", flowPathId)));
+        PathId oppositePathId = flow.getOppositePathId(flowPathId)
+                .orElseThrow(() -> new FlowProcessingException(ErrorType.NOT_FOUND,
+                        format("Opposite flow path id for path %s not found", flowPathId)));
+        FlowPath oppositePath = flow.getPath(oppositePathId).orElse(null);
+        RequestedFlowMirrorPoint mirrorPoint = stateMachine.getRequestedFlowMirrorPoint();
+
+        CommandContext context = stateMachine.getCommandContext();
+        if (mirrorPoint.getMirrorPointSwitchId().equals(path.getSrcSwitchId())) {
+            return new ArrayList<>(commandBuilder
+                    .buildIngressOnlyOneDirection(context, flow, path, oppositePath,
```

Comment:
This ingress/egress/transit segment kinds selection should be done inside commandBuilder i.e. it `commanBuilder` responsibility/layer. If it will be kept in current state it will be hard to make mirror path comands in future changes.